### DISCUSSION TITLE
Style Guide for Mojo Code. Zen of Mojo

### DISCUSSION THREAD

**Discussion by ego** [modular/modular#141] [General]
Maybe it is not so important, but actually does, what is the **convention for naming** in a languages? 
Function/methods **CamelCase/snake_case**?

In the Python world, there are clear rules [PEP8](https://peps.python.org/pep-0008/) about when to `use snake_case/CamelCase`. This makes the code elegant and easy to read.

Definitely language should has built-in tools for format and lint code (`rustfmt/swiftformat`).

The question is when to have such tools. We can wait a bit, it's not critical, even the community can write their own tools, but it can create some chaos. As with Python, there are many formatting tools (Black/pylint/yapf) and code guidelines ([Google](https://google.github.io/styleguide/pyguide.html), PEP8, etc.).

With tools like `fmt`, a language can easily build and change its keywords, syntax, other things, and devs can easily reformat their code to accommodate the new changes without pain. Just like python2 smoothly migrates to python3 with [2to3](https://docs.python.org/3/library/2to3.html).

Open question is about "Zen of Mojo".

**Comment by lattner** [comment]
We are a member of the Python family, so following PEP8 is the right thing to do.

**Reply by ego** [reply]
Jupyter notebooks is in a private space, so we can't create a PR for docs.
But it would be useful to be able to create a PR with some fixes/corrections/typos while reading the documentation.

Like (2 spaces before #, test function prefix naming, etc)

```
def pairTest() -> Bool:
    let p = MyPair(1, 2)
    # Uncomment to see an error:
    # return p < 4  # gives a compile time error
    return True
```

to

```
def test_pair() -> Bool:
    let p = MyPair(1, 2)
    # Uncomment to see an error:
    # return p < 4  # gives a compile time error
    return True


print(test_pair())
```
